### ISSUE TITLE
cephfs in GKE: PV / PVCs not being deleted

### ISSUE THREAD

**Issue by Mattsky** [kubernetes-incubator/external-storage#860]
PVs and PVCs are being created properly using the provisioner, but once the hosting namespace is deleted, the cleanup fails. Logs follow:

`E0717 01:49:33.363126       1 cephfs-provisioner.go:261] failed to delete share "kubernetes-dynamic-pvc-ccaa8dc9-8959-11e8-8e61-625c2d54d0bc" for "kubernetes-dynamic-user-ccaa8e64-8959-11e8-8e61-625c2d54d0bc", err: exit status 1, output: Traceback (most recent call last):
  File "/usr/local/bin/cephfs_provisioner", line 336, in <module>
    main()
  File "/usr/local/bin/cephfs_provisioner", line 332, in main
    cephfs.delete_share(share, user)
  File "/usr/local/bin/cephfs_provisioner", line 291, in delete_share
    self._deauthorize(volume_path, user_id)
  File "/usr/local/bin/cephfs_provisioner", line 242, in _deauthorize
    pool_name = self.volume_client._get_ancestor_xattr(path, "ceph.dir.layout.pool")
  File "/lib/python2.7/site-packages/ceph_volume_client.py", line 756, in _get_ancestor_xattr
    result = self.fs.getxattr(path, attr)
  File "cephfs.pyx", line 954, in cephfs.LibCephFS.getxattr (/home/jenkins-build/build/workspace/ceph-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/13.2.0/rpm/el7/BUILD/ceph-13.2.0/build/src/pybind/cephfs/pyrex/cephfs.c:10083)
cephfs.ObjectNotFound: [Errno 2] error in getxattr
E0717 01:49:33.363179       1 controller.go:1314] Deletion of volume "pvc-cca339bb-8959-11e8-ad6a-42010a80000e" failed: exit status 1`

This is running the v1.1.0-k8s1.10 image in GKE with nodes running version 1.10.5-gke.0, and Ceph version 12.2.5 (cad919881333ac92274171586c827e01f554a70a) luminous. RBAC is in place and the provisioner was deployed using the RBAC deployment files. Not seeing any clues in the logs on the monitor. Anyone seen anything like this before?

**Comment by Mattsky** [comment]
Followup: did some checking, it seems this release is geared for Mimic, where the cluster is Luminous. Going to try upgrading the cluster and see if that resolves the issue.

**Comment by Mattsky** [comment]
Cluster's upgraded to Mimic, and the problem persists.

**Comment by Mattsky** [comment]
I've tried getting a shell into the provisioner container, and have confirmed it can talk to the cluster. I've also tried deleting the PV / PVC only before deleting the namespace and get the same error as originally:

`[root@cephfs-provisioner-6bfb6d6467-x2kmw ceph]# CEPH_AUTH_KEY=<KEY_HERE> CEPH_AUTH_ID=<ADMIN_USER_HERE> CEPH_MON=<MONITOR_IP> python /usr/local/bin/cephfs_provisioner -r -n kubernetes-dynamic-pvc-54a71e66-8a02-11e8-8e61-625c2d54d0bc -u kubernetes-dynamic-user-54a720ab-8a02-11e8-8e61-625c2d54d0bc
Traceback (most recent call last):
  File "/usr/local/bin/cephfs_provisioner", line 336, in <module>
    main()
  File "/usr/local/bin/cephfs_provisioner", line 332, in main
    cephfs.delete_share(share, user)
  File "/usr/local/bin/cephfs_provisioner", line 291, in delete_share
    self._deauthorize(volume_path, user_id)
  File "/usr/local/bin/cephfs_provisioner", line 242, in _deauthorize
    pool_name = self.volume_client._get_ancestor_xattr(path, "ceph.dir.layout.pool")
  File "/usr/lib/python2.7/site-packages/ceph_volume_client.py", line 756, in _get_ancestor_xattr
    result = self.fs.getxattr(path, attr)
  File "cephfs.pyx", line 954, in cephfs.LibCephFS.getxattr (/home/jenkins-build/build/workspace/ceph-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/13.2.0/rpm/el7/BUILD/ceph-13.2.0/build/src/pybind/cephfs/pyrex/cephfs.c:10083)
cephfs.ObjectNotFound: [Errno 2] error in getxattr`

I half-wondered if it could've been a race condition (something required being deleted in the namespace before the PV/PVC deletion op was run?) but I guess not.

**Comment by Mattsky** [comment]
More info: this has only started happening since updating the container image from:

`quay.io/external_storage/cephfs-provisioner@sha256:d3be97e4f67632b5d3daed7fb38c5dbb160eeae34fc9fa79a03590a9bf7cc0d3`

**Comment by Mattsky** [comment]
Another update - I reverted the container back to the version I listed in the previous comment and tried the ops again. This time, I got the following:

`$ k logs -f cephfs-provisioner-78c9b8f54c-d4zz4 -n cephfs
I0718 21:32:44.120905       1 cephfs-provisioner.go:403] Creating CephFS provisioner ceph.com/cephfs with identity: cephfs-provisioner-1, secret namespace:
I0718 21:32:44.138002       1 controller.go:548] Starting provisioner controller 1b57c1fa-8ad2-11e8-80e6-6eb9a45be078!
I0718 21:34:50.089471       1 controller.go:1264] scheduleOperation[delete-pvc-9659f1f3-8acb-11e8-ad6a-42010a80000e[96b46913-8acb-11e8-ad6a-42010a80000e]]
I0718 21:34:59.245432       1 controller.go:1264] scheduleOperation[delete-pvc-9659f1f3-8acb-11e8-ad6a-42010a80000e[96b46913-8acb-11e8-ad6a-42010a80000e]]
E0718 21:35:04.240964       1 cephfs-provisioner.go:273] Cephfs Provisioner: delete secret failed, err: secrets "ceph-kubernetes-dynamic-user-966055c3-8acb-11e8-830b-aee5241b11d7-secret" not found
E0718 21:35:04.241393       1 controller.go:1235] Deletion of volume "pvc-9659f1f3-8acb-11e8-ad6a-42010a80000e" failed: failed to delete secret
E0718 21:35:04.243489       1 goroutinemap.go:165] Operation for "delete-pvc-9659f1f3-8acb-11e8-ad6a-42010a80000e[96b46913-8acb-11e8-ad6a-42010a80000e]" failed. No retries permitted until 2018-07-18 21:35:04.743392067 +0000 UTC m=+140.754518742 (durationBeforeRetry 500ms). Error: "failed to delete secret"
I0718 21:35:14.245831       1 controller.go:1264] scheduleOperation[delete-pvc-9659f1f3-8acb-11e8-ad6a-42010a80000e[96b46913-8acb-11e8-ad6a-42010a80000e]]
E0718 21:35:14.619070       1 cephfs-provisioner.go:261] failed to delete share "kubernetes-dynamic-pvc-966054d5-8acb-11e8-830b-aee5241b11d7" for "kubernetes-dynamic-user-966055c3-8acb-11e8-830b-aee5241b11d7"`

I think maybe it is a race condition after all, going by the `delete secret failed` comment above. Insight definitely appreciated. 😄

**Comment by wongma7** [comment]
Have you configured the class parameter `adminSecretNamespace`? https://github.com/kubernetes-incubator/external-storage/blob/master/ceph/cephfs/example/class.yaml#L10 IIRC it was implemented to isolate secrets from user's namespace because they shouldn't have access to them & in case they delete them like here.

**Comment by Mattsky** [comment]
Yep:

`
$ k get sc cephfs -o yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: | BLAH
  creationTimestamp: 2018-04-26T17:27:15Z
  name: cephfs
  resourceVersion: "33017023"
  selfLink: /apis/storage.k8s.io/v1/storageclasses/cephfs
  uid: 0ff4a76c-4977-11e8-ad73-42010a800046
parameters:
  adminId: ADMIN_ID
  adminSecretName: ceph-secret-admin
  adminSecretNamespace: NAMESPACE
  monitors: MONITOR_IP:6789
provisioner: ceph.com/cephfs
reclaimPolicy: Delete
volumeBindingMode: Immediate
`

The admin secret's in a different namespace and is present.

**Comment by Mattsky** [comment]
I've done a little more investigation and restarted the pod just to see what happened. The following output seems to indicate the secret namespace is blank:

```
$ k logs cephfs-provisioner-b597cc69c-5hkwg -n cephfs
I0725 16:17:58.849760       1 cephfs-provisioner.go:403] Creating CephFS provisioner ceph.com/cephfs with identity: cephfs-provisioner-1, secret namespace:
I0725 16:17:59.088369       1 controller.go:615] Starting provisioner controller 4bc355b1-9026-11e8-be56-1e6f5d05d357!
I0725 16:17:59.088466       1 controller.go:652] Started provisioner controller 4bc355b1-9026-11e8-be56-1e6f5d05d357!
```

I'm assuming that's the secret namespace for each dynamic user secret as opposed to the admin one? I'm going to try setting this in the container env and see if this sidesteps the issue.

**Comment by Mattsky** [comment]
After trying that, the provisioner still doesn't clean up the PV after the namespace is deleted. No errors are thrown, at least.

**Comment by zhoubofsy** [comment]
check `ceph.dir.layout.pool` attr:
    `getfattr -n ceph.dir.layout.pool <path>`

**Comment by zhoubofsy** [comment]
1. mount cephfs to `/mnt`
2. check ceph.dir.layout.pool attr
`getfattr -n ceph.dir.layout.pool /mnt`
3. if no such attribute
`setfattr -n ceph.dir.layout.pool -v "cephfs_data" /mnt`

**Comment by xriser** [comment]
same issue
`E0322 13:02:25.821003       1 cephfs-provisioner.go:268] failed to delete share "data-elasticsearch-elasticsearch-master-0" for "k8s.efk.data-elasticsearch-elasticsearch-master-0", err: exit status 1, output: Traceback (most recent call last):
  File "/usr/local/bin/cephfs_provisioner", line 364, in <module>
    main()
  File "/usr/local/bin/cephfs_provisioner", line 360, in main
    cephfs.delete_share(share, user)
  File "/usr/local/bin/cephfs_provisioner", line 319, in delete_share
    self._deauthorize(volume_path, user_id)
  File "/usr/local/bin/cephfs_provisioner", line 260, in _deauthorize
    pool_name = self.volume_client._get_ancestor_xattr(path, "ceph.dir.layout.pool")
  File "/lib/python2.7/site-packages/ceph_volume_client.py", line 756, in _get_ancestor_xattr
    result = self.fs.getxattr(path, attr)
  File "cephfs.pyx", line 954, in cephfs.LibCephFS.getxattr (/home/jenkins-build/build/workspace/ceph-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/13.2.1/rpm/el7/BUILD/ceph-13.2.1/build/src/pybind/cephfs/pyrex/cephfs.c:10083)
cephfs.ObjectNotFound: [Errno 2] error in getxattr
E0322 13:02:25.821248       1 controller.go:1181] delete "pvc-047c2c0f-30a7-4e73-acfc-a79bf4460dcd": volume deletion failed: exit status 1
I0322 13:02:25.821665       1 event.go:221] Event(v1.ObjectReference{Kind:"PersistentVolume", Namespace:"", Name:"pvc-047c2c0f-30a7-4e73-acfc-a79bf4460dcd", UID:"57f96660-0ee2-4d5b-96d6-a82837049561", APIVersion:"v1", ResourceVersion:"440515", FieldPath:""}): type: 'Warning' reason: 'VolumeFailedDelete' exit status 1
`
Tried mimic and nautilus, rebuilt provisioner docker image to the latest code, tried different images, checked fattr as @zhoubofsy mentioned, nothing helped
Also checked secrets and so on.

The issue, as follows: pvc provisions well, client secret creates in cephfs namespace as well.
Then when I deleted pvc - it deleted, pv still remaining with Released status, the data from pvc actually remaining in the ceph storage, the client secret does not delete, and the error in the provisioner pod.
Then I can delete pv and user secret by hands - it is ok. But the data in the ceph storage still remaining.
Then if I create pvc with the same name - it will be with the previously stored data.
